pkg/3scale/amp/operator/upgrade.go

```diff
@@ -614,6 +628,97 @@ func (u *UpgradeApiManager) upgradeSystemMasterApicastSecret() (reconcile.Result
 	return reconcile.Result{}, nil
 }
 
+func (u *UpgradeApiManager) upgradePodTemplateLabels() (reconcile.Result, error) {
+	apicast, err := Apicast(u.apiManager)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	backend, err := Backend(u.apiManager, u.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	zync, err := Zync(u.apiManager, u.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	memcached, err := Memcached(u.apiManager)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	system, err := System(u.apiManager, u.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	deploymentConfigs := []*appsv1.DeploymentConfig{
+		apicast.StagingDeploymentConfig(),
+		apicast.ProductionDeploymentConfig(),
+		backend.ListenerDeploymentConfig(),
+		backend.WorkerDeploymentConfig(),
+		backend.CronDeploymentConfig(),
+		zync.DeploymentConfig(),
+		zync.QueDeploymentConfig(),
+		zync.DatabaseDeploymentConfig(),
+		memcached.DeploymentConfig(),
+		system.AppDeploymentConfig(),
+		system.SidekiqDeploymentConfig(),
+		system.SphinxDeploymentConfig(),
+	}
+
+	if !u.apiManager.IsExternalDatabaseEnabled() {
+		redis, err := Redis(u.apiManager)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		deploymentConfigs = append(deploymentConfigs, redis.SystemDeploymentConfig())
+		deploymentConfigs = append(deploymentConfigs, redis.BackendDeploymentConfig())
+	}
+
+	if u.apiManager.IsSystemPostgreSQLEnabled() {
+		systemPostgreSQL, err := SystemPostgreSQL(u.apiManager, u.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		deploymentConfigs = append(deploymentConfigs, systemPostgreSQL.DeploymentConfig())
+	} else {
+		systemMySQL, err := SystemMySQL(u.apiManager, u.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		deploymentConfigs = append(deploymentConfigs, systemMySQL.DeploymentConfig())
+	}
+
+	for _, desired := range deploymentConfigs {
+		err = u.ensurePodTemplateLabels(desired)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
```

Comment:
It would probably be a good idea when changes happen so the upgrade part is reevaluated again to ensure changes have been applied and there are no new changes, before finishing the upgrade completely. We could do it at the end of the for loop if there's been a change or at each iteration if there's been a change.

What do you think?

Reply:
done